test/cpp/util/channelz_sampler.cc

```diff
@@ -0,0 +1,289 @@
+    std::queue<grpc::channelz::v1::Subchannel>& subchannel_queue) {
+  std::cout << "    Subchannel " << GetSubchannelID(subchannel)
+            << " descendence - ";
+  if (subchannel.channel_ref_size() > 0) {
+    std::cout << "channel: ";
+    for (auto& i : subchannel.channel_ref()) {
+      int64_t ch_id = i.channel_id();
+      std::cout << ch_id << " ";
+      GetChannelRPC(ch_id, channelz_stub, channel_queue);
+    }
+  }
+  if (subchannel.subchannel_ref_size() > 0) {
+    std::cout << "subchannel: ";
+    for (auto& i : subchannel.subchannel_ref()) {
+      int64_t subch_id = i.subchannel_id();
+      std::cout << subch_id << " ";
+      GetSubchannelRPC(subch_id, channelz_stub, subchannel_queue);
+    }
+  }
+  if (subchannel.socket_ref_size() > 0) {
+    std::cout << "socket: ";
+    for (auto& i : subchannel.socket_ref()) {
+      int64_t so_id = i.socket_id();
+      std::cout << so_id << " ";
+    }
+  }
+  std::cout << std::endl;
+}
+
+int main(int argc, char** argv) {
+  // make sure flags can be used
+  grpc::testing::InitTest(&argc, &argv, true);
+  std::cout << "server address: " << FLAGS_server_address << std::endl;
+  std::cout << "custom credentials type: " << FLAGS_custom_credentials_type
+            << std::endl;
+
+  // create a channelz client
+  ::grpc::channelz::experimental::InitChannelzService();
+  grpc::ChannelArguments channel_args;
+  std::shared_ptr<grpc::ChannelCredentials> channel_creds =
+      grpc::testing::GetCredentialsProvider()->GetChannelCredentials(
+          FLAGS_custom_credentials_type, &channel_args);
+  std::shared_ptr<grpc::Channel> channel =
+      CreateChannel(FLAGS_server_address, channel_creds);
+  std::unique_ptr<grpc::channelz::v1::Channelz::Stub> channelz_stub =
+      grpc::channelz::v1::Channelz::NewStub(channel);
+
+  // Server side code
+  // Get all servers by calling GetServers
+  GetServersRequest get_server_request;
+  GetServersResponse get_server_response;
+  ClientContext get_server_context;
+  std::vector<grpc::channelz::v1::Server> top_servers;
+  int64_t server_start_id = 0;
+  while (true) {
+    get_server_request.set_start_server_id(server_start_id);
+    Status status = channelz_stub->GetServers(
+        &get_server_context, get_server_request, &get_server_response);
+    if (!status.ok()) {
+      std::cout << "Channelz failed: " << status.error_message() << std::endl;
+      return 1;
+    }
+    for (auto& i : get_server_response.server()) {
+      top_servers.push_back(i);
+    }
+    if (!get_server_response.end()) {
+      server_start_id = GetServerID(top_servers.back()) + 1;
+    } else {
+      break;
+    }
+  }
+  std::cout << "Number of servers = " << top_servers.size() << std::endl;
+
+  // print out sockets of each server
+  for (auto& i : top_servers) {
+    std::cout << "Server " << GetServerID(i) << " listen_socket: ";
+    for (auto& j : i.listen_socket()) {
+      std::cout << j.socket_id() << " ";
```

Comment:
It would be more interesting for debug purposes to see the entire `Server` data structure.

Can we use the protobuf text format APIs to dump the entire `Server` objects (https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.text_format)

Reply:
done